### PR TITLE
feat: onboarding global dashboard, multiple onboarding sequence

### LINKED ISSUE #5161: Global onboarding: Add to dashboard a selectable list of onboarding sequences that need setting up
### Is your enhancement related to a problem? Please describe

Part of the epic: https://github.com/containers/podman-desktop/issues/5045

Implement the "global dashboard" that will list all the providers that have onboarding sequences available. 

Give the option to run the setup if the onboarding sequence has not been setup / ran yet / it is detected that you need to set it up.

### Describe the solution you'd like

Implement below mockup

image

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### LINKED ISSUE #5162: Global onboarding: Implement skip/cancel entire setup / skip onboarding.
### Is your enhancement related to a problem? Please describe

Part of the epic #5045 

Implement a way to skip certain parts of the onboarding (for example, totally skip Podman, or Compose, etc.).

### Describe the solution you'd like

Implement the below mockup:

image

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### LINKED ISSUE #5163: Global onboarding: Pressing "Esc" should cancel the entire sequence
### Is your feature request related to a problem? Please describe

Part of the epic #5045 

Pressing ESC should cancel the entire sequence / skip it.

### Describe the solution you'd like

Pressing ESC cancels / skips the entirety of it.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by cdrage**
feat: onboarding global dashboard, multiple onboarding sequence

### What does this PR do?

* Adds a "global onboarding" dashboard to the welcome page
* Modifies Onboarding.svelte to allow being able to run multiple
  onboarding sequences.
* When multiple onboarding sequences are to be completed, onboarding
  goes "full screen" and hides the status bar / side bar.

### Screenshot / video of UI

https://github.com/containers/podman-desktop/assets/6422176/f7ee16ee-fd0e-40d9-b11c-7c9f3ae70070

### What issues does this PR fix or reference?

Closes https://github.com/containers/podman-desktop/issues/5162
Closes https://github.com/containers/podman-desktop/issues/5161
Closes https://github.com/containers/podman-desktop/issues/5163

### How to test this PR?

1. Remove `welcome.version` from ~/.local/share/containers/podman-desktop/configuration/settings.json
so you can access the welcome screen
2. Click on multiple onboarding selections
3. Go through the onboarding sequence

Signed-off-by: Charlie Drage <charlie@charliedrage.com>

### CODE REVIEW

**Comment by cdrage** [comment]
Screenshot 2023-12-14 at 3 35 24 PM

I'd like some suggestions on how we could improve showing the onboarding. Right now I am showing anything that has onboarding enabled, not necessarily "needing" setting up.

This also would be removing the "featured" list..

So would we add featured items as well and simply not have the checkbox available?

@mairin @deboer-tim

**Comment by cdrage** [comment]
TODO:

- [ ] Add tests
- [ ] Finalize UI

**Comment by benoitf** [comment]
hello, shouldn't the checkboxes be purple and smaller like the one used for telemetry just below the cards ?

and is that the default should be checked ?

**Comment by cdrage** [comment]
> hello, shouldn't the checkboxes be purple and smaller like the one used for telemetry just below the cards ?
> 
> and is that the default should be checked ?

288557538-fe7aca3f-e998-4102-a76a-fccc073c8ccf (1)

I'm going by the mockup so that's why they are larger and not clicked by default.

My UI & integration questions are:

Should I be removing the "featured list" as now we are using onboarding?

In the mockup is shows "featured" labels but you can check mark them even though it's not onboarding, what does that mean?

Should we be showing ALL the onboarding even if it doesn't require setting up? Or else it would be very empty vs the featured list. Each onboarding process has a "summary" page at the end, so we would show that? Or do we just mark it as grayed out / already setup?

**Comment by mairin** [comment]
> I'm going by the mockup so that's why they are larger and not clicked by default.

👍🏻 

> Should I be removing the "featured list" as now we are using onboarding? 
> In the mockup is shows "featured" labels but you can check mark them even though it's not onboarding, what does that mean?

Ah this is a good question. The mockup was created with the assumption every extension shown has an onboarding flow. This is clearly not the case now. So I would only show the extensions that have an onboarding flow available and drop the featured list. (We show the featured list on the dashboard anyway, so they get some good promotion to users without being in the welcome screen.)

> Should we be showing ALL the onboarding even if it doesn't require setting up? Or else it would be very empty vs the featured list. Each onboarding process has a "summary" page at the end, so we would show that? Or do we just mark it as grayed out / already setup?

So for Linux users this question is of particular interest wrt Podman... because if they're using a RHT family distro anyway, they'll have podman already installed. But they may not have set up configuration specific to Podman Desktop in respect to that - I guess the main one is does it auto start or not. So my initial thoughts are:

- I cannot think of any cases where the user going through the welcome screen - which should only appear on the first run of the application - would actually have the extension + software already installed except for: (1) Podman on Linux (2) manually installing podman-machine on Mac/Win (3) manually installing crc on mac/win/linux

- For the Linux case, I'd want to see the podman extension options even if I already have podman installed, because I want to see what preferences I can set wrt it.

- For the manually-installed podman-machine / crc... I think going through the flow lets me see whether or not it detected my manually installed thing and tweak configuration, so it's not a problem to go through it again.

So I would err on the side of letting them go through it again. I think if we detect an installation, we could flag that with one of those little taggy things at the bottom of the card, "Installation detected" or something like that. There may even be a mockup for this in the penpot file, I'll dig around and look, because it feels familiar.

**Comment by mairin** [comment]
> So I would err on the side of letting them go through it again. I think if we detect an installation, we could flag that with one of those little taggy things at the bottom of the card, "Installation detected" or something like that. There may even be a mockup for this in the penpot file, I'll dig around and look, because it feels familiar.

Ah yeh I went back through the mockup file. The "detected" badge is meant to show we detected that you had the thing installed.

**Comment by slemeur** [comment]
Thanks @cdrage this is shaping together. 

To your question:
> Should I be removing the "featured list" as now we are using onboarding?

I assume you mean the featured extensions that are displayed on the dashboard. In that's really the ones you meant, then no - we should keep them for the time being. 
We can refactor the dashboard in a separate effort.

**Comment by mairin** [comment]
Will talk about this at the UX call tomorrow!

**Comment by mairin** [comment]
Quick notes from the UX call yesterday:

- Keep the featured extensions at the bottom, but use the small / non-clickable version. Maybe drop the button-looking background off of them so they don't look like clickable buttons since they're not. 
- Have some logic in the featured extensions area to de-duplicate what's offered via onboarding checkbox options at the top.
- Remove the podman option if on linux for the featured extension area.

**Comment by cdrage** [comment]
Ready for review!

Changes:
- Updated UI to still show featured list, but remove the ones that already have onboarding
- Added a *a lot* of tests



https://github.com/containers/podman-desktop/assets/6422176/1b41990a-51d0-4d72-b381-7739b86fcb5a

**Comment by cdrage** [comment]
@benoitf @dgolovin @jeffmaury Ready for review 🎉

**Comment by odockal** [comment]
@cdrage Please also map this PR or issue addressed here into this [epic](https://github.com/containers/podman-desktop/issues/5491), to be sure we are counting in QE with this feature to be covered by the tests in the future.

**Review by lstocchi — COMMENTED** [review]
Unfortunately it doesn't work for me. The podman onboarding is broken, even if i start it from the resources page the page is empty - no errors in the console :/ (also tried to run it from the main and works as expected, so i guess it is some update from here). I'm using Windows.

EDIT: to be precise, during the global onboarding, the podman onboarding get stuck at the installing podman step (there should be a dialog asking for a confirmation to install it and it does not appear). When i start the podman onboarding from the resources page, i don't see anything. All black

Rgearding the UI, not sure if it's correct but i see a light grey row  between the top toolbar and the onboarding header. Is that expected?

image

**Comment by cdrage** [comment]
> Unfortunately it doesn't work for me. The podman onboarding is broken, even if i start it from the resources page the page is empty - no errors in the console :/ (also tried to run it from the main and works as expected, so i guess it is some update from here). I'm using Windows.
> 
> EDIT: to be precise, during the global onboarding, the podman onboarding get stuck at the installing podman step (there should be a dialog asking for a confirmation to install it and it does not appear). When i start the podman onboarding from the resources page, i don't see anything. All black
> 
> Rgearding the UI, not sure if it's correct but i see a light grey row between the top toolbar and the onboarding header. Is that expected?
> 
> image

Any way to take a video? Just trying to reproduce it and have been unable to (even uninstalling podman and trying it again with a fresh onboarding).

**Comment by cdrage** [comment]
> Unfortunately it doesn't work for me. The podman onboarding is broken, even if i start it from the resources page the page is empty - no errors in the console :/ (also tried to run it from the main and works as expected, so i guess it is some update from here). I'm using Windows.
> 
> EDIT: to be precise, during the global onboarding, the podman onboarding get stuck at the installing podman step (there should be a dialog asking for a confirmation to install it and it does not appear). When i start the podman onboarding from the resources page, i don't see anything. All black
> 
> Rgearding the UI, not sure if it's correct but i see a light grey row between the top toolbar and the onboarding header. Is that expected?
> 
> image

I've just rebased against master and I'm trying to reproduce it on Windows and unfortunately I'm unable to get the black screen.

**Comment by cdrage** [comment]
https://github.com/containers/podman-desktop/assets/6422176/dbd00a66-37b0-449e-a962-21dc5f007f4b

@lstocchi Indeed it was a z-value error! It had me scratching my head why it wasn't appearing, and it was just getting hidden.

I've updated the PR with the small minor change and it's ready for a review.

@lstocchi @jeffmaury @dgolovin

File: packages/renderer/src/lib/welcome/WelcomePage.svelte [hunk 1]
```diff
@@ -3,32 +3,61 @@ import bgImage from './background.png';
...
+
+  onboardingsUnsubscribe = onboardingList.subscribe(value => {
+    // Add "selected" property to each provider and add to onboardingEnabledProviders
+    onboardingProviders = value.map(provider => {
+      return {
+        ...provider,
+        selected: false,
+      };
+    });
+    console.log(onboardingProviders);
```
Comment (lstocchi) [inline packages/renderer/src/lib/welcome/WelcomePage.svelte]: to be removed?

**Review by lstocchi — COMMENTED** [review]
I still see a black page when i skip the global onboarding and i select the podman one from the resources page (or compose).

After investigating a bit, the problem is in this function
https://github.com/containers/podman-desktop/blob/46d8ea646eb0eabf266a74bc16b046786d668cf1/packages/renderer/src/lib/onboarding/onboarding-utils.ts#L44-L57

When the user skips the global onboarding you set all steps to "skipped" in the right way. However, by doing this, the method above will set the entire onboarding to complete when the status of the last step gets updated (to skipped). 
We do not do any check here, so no matter which status the last step has. If the last step gets updated we mark the onboarding as completed. For this reason, if you then try to open the onboarding from the resources page you see nothing bc for desktop the onboarding has been completed before.

To avoid it you just need to add a check before updating the entire onboarding status like 

`if (lastCompletedStep?.id === step.id && status === STATUS_COMPLETED) {`

this way the onboarding will be marked as complete only if the last step is completed.

File: packages/renderer/src/lib/onboarding/Onboarding.svelte [hunk 2]
```diff
@@ -71,7 +73,7 @@ let executedCommands: string[] = [];
 let telemetrySession = new OnboardingTelemetrySession();
 
 /*
-$: enableNextButton = false;*/
+  $: enableNextButton = false;*/
```
Comment (lstocchi) [inline packages/renderer/src/lib/onboarding/Onboarding.svelte]: to be removed?

**Comment by cdrage** [comment]
@lstocchi Thanks again for testing.

I've added the new check and I'm able to "skip" the kubectl part now and still install it, do you mind reviewing again? Thanks for being patient through all of this.

https://github.com/containers/podman-desktop/assets/6422176/227ae3f8-5f57-4c9b-be4e-e9dec5768e6b

**Review by lstocchi — APPROVED** [review]
LGTM. Thanks @cdrage , great job 🚀

**Comment by jeffmaury** [comment]
If I don't have Compose installed (for example) and I go to resource pages, it tells me Set up.
Can't we have the Compose card selected in that case in the onboard screen ?

**Comment by cdrage** [comment]
> If I don't have Compose installed (for example) and I go to resource pages, it tells me Set up. Can't we have the Compose card selected in that case in the onboard screen ?

That is intentional with the design not to have it check-boxed as users may not want to initially set up the extension.

I also kept encountering this bug: https://github.com/containers/podman-desktop/issues/5249 which would make loading context on the initial welcome screen not very stable.

**Comment by slemeur** [comment]
> That is intentional with the design not to have it check-boxed as users may not want to initially set up the extension.

If it's not setup, I think we should default its installation / configuration. Those who don't want it, can uncheck the box.
The users who already have it installed, will not need this step as the compatibility mode will cover it. (in this case, it should be skipped)
We still do have quite a few users who are reporting struggling with the configuration of compose - hopefully the onboarding sequence will make things much clearer and simpler for the end-users

**Comment by cdrage** [comment]
> > That is intentional with the design not to have it check-boxed as users may not want to initially set up the extension.
> 
> If it's not setup, I think we should default its installation / configuration. Those who don't want it, can uncheck the box. The users who already have it installed, will not need this step as the compatibility mode will cover it. (in this case, it should be skipped) We still do have quite a few users who are reporting struggling with the configuration of compose - hopefully the onboarding sequence will make things much clearer and simpler for the end-users

Managed to get this to work!

I've updated the PR to include retrieving the current providers (in order to get the status), as well as the global context and pre-select the ones that have enablement.

I've added it as a separate PR and implemented the requested changes @slemeur 


https://github.com/containers/podman-desktop/assets/6422176/098a601d-8d25-4b16-b2f9-c3f896365c57

@lstocchi do you mind testing this again when you get the chance? As you already have a Windows setup all ready to go 🎉

**Review by jeffmaury — CHANGES_REQUESTED** [review]
Found a blocker: for the Podman onboarding, once Podman is installed, I need to create a Podman machine but there is no way to scroll in the page

image

Also if Podman is installed but no Podman machine I think we should have the Podman onboarding selected in the onboarding page

**Comment by lstocchi** [comment]
Tested again, same as Jeff

**Comment by cdrage** [comment]
@jeffmaury @lstocchi Looks like I forgot to add the overflow-y-auto.

Re-added it and now you can scroll correctly:

https://github.com/containers/podman-desktop/assets/6422176/0b24cce3-523f-4fcb-8e15-e817ecc8c56d


To answer your question on:

"Also if Podman is installed but no Podman machine I think we should have the Podman onboarding selected in the onboarding page"

We use the global context  / onbarding onWhen clauses and would have to implement provider-specific support, see how the code works with showing "Create new ..." vs "Setup ...": https://github.com/cdrage/podman-desktop/blob/3d81f974ffef98015aa8957073815a103ca4305b/packages/renderer/src/lib/preferences/PreferencesResourcesRendering.svelte#L402

As this PR is getting too big, let's do it in a follow up PR?

**Comment by jeffmaury** [comment]
The Podman onboarding is broken for me in the autostart page

image

**Comment by cdrage** [comment]
> The Podman onboarding is broken for me in the autostart page
> 
> image

ACK. Looks like it's a bug with the styling after rebasing against the new changes.

**Review by jeffmaury — APPROVED** [review]
LGTM

File: packages/main/src/plugin/onboarding-registry.ts [hunk 3]
```diff
@@ -54,9 +54,14 @@ export class OnboardingRegistry {
     this.checkIdsReadability(extension, onboarding);
     //TODO we need to check the onboarding has a valid schema. contains atleast a step and substep
     this.convertImages(extension, onboarding);
+
     return {
       ...onboarding,
       extension: extension.id,
+      name: extension.name,
+      displayName: extension.manifest?.displayName || extension.name,
```
Comment (benoitf) [inline packages/main/src/plugin/onboarding-registry.ts]:
```suggestion
      displayName: extension.manifest?.displayName ?? extension.name,
```

File: packages/main/src/plugin/onboarding-registry.ts [hunk 4]
```diff
@@ -54,9 +54,14 @@ export class OnboardingRegistry {
...
+      description: extension.manifest?.description || '',
```
Comment (benoitf) [inline packages/main/src/plugin/onboarding-registry.ts]:
```suggestion
      description: extension.manifest?.description ?? '',
```

File: packages/main/src/plugin/onboarding-registry.ts [hunk 5]
```diff
@@ -54,9 +54,14 @@ export class OnboardingRegistry {
...
+      icon: onboarding.media?.path || '',
```
Comment (benoitf) [inline packages/main/src/plugin/onboarding-registry.ts]:
```suggestion
      displayName: extension.manifest?.displayName || extension.name,
      description: extension.manifest?.description || '',
      icon: onboarding.media?.path ?? '',
```

**Review by benoitf — CHANGES_REQUESTED** [review]
The first screen people will see is broken 
image

we had in the past something 'slick' here we have a vertical scrollbar and I don't see the Podman Desktop logo correctly
Also we don't see the line
image
without scrolling

IMHO this first screen should be displayed without any scrolling and we should see every item. This is the first thing users will see


Then, why the checkboxes are not checked by default for the onboarding ?

Clicking on a checkbox, it's moving the UI when it's checked
https://github.com/containers/podman-desktop/assets/436777/dd718e8a-b425-456f-b148-8a16b0a44e96
Not sure why the checkboxes are blue and not purple ?

On the UI, I have the feeling that I have a toggle to switch from one step to another step
image

I think I could click here to go back to previous steps




For Podman onboarding, I see the screen
image

so I have the feeling that everything is installed but if I click next I have
image

I don't see what this screen is bringing I would expected to not see this screen as it was saying that it was already installed on the previous screen.

**Comment by cdrage** [comment]
> The first screen people will see is broken image
> 
> we had in the past something 'slick' here we have a vertical scrollbar and I don't see the Podman Desktop logo correctly Also we don't see the line image without scrolling
> 
> IMHO this first screen should be displayed without any scrolling and we should see every item. This is the first thing users will see
> 
> Then, why the checkboxes are not checked by default for the onboarding ?
> 
> Clicking on a checkbox, it's moving the UI when it's checked https://github.com/containers/podman-desktop/assets/436777/dd718e8a-b425-456f-b148-8a16b0a44e96 Not sure why the checkboxes are blue and not purple ?
> 
> On the UI, I have the feeling that I have a toggle to switch from one step to another step image
> 
> I think I could click here to go back to previous steps
> 
> For Podman onboarding, I see the screen image
> 
> so I have the feeling that everything is installed but if I click next I have image
> 
> I don't see what this screen is bringing I would expected to not see this screen as it was saying that it was already installed on the previous screen.

Screenshot 2024-03-04 at 3 07 48 PM

With #6268 we got some space back now.

The above implementation of dashboard was done via the UI mockup.

To answer your questions:

1. The purple "dots" are from the UI mockup. I'm unsure what we should do with regards to that / make it better, as they do look "clickable"
2. Also unsure regarding creating space, as we concluded we needed the "featured" section too, should we just remove the "Configure these and more under Settings" text?
3. Checkboxes are only checked if it's needing Setup, similar to the "Setup ..." button appearing under Extensions in settings. However, you CAN click on it again to go through the process such as the "extra" options like toggling auto start / on and off.


EDIT: I have opened up https://github.com/containers/podman-desktop/pull/6270 as well to address the "Podman successfully setup" screen.

**Comment by benoitf** [comment]
@cdrage there was agreement to get rid of featured extensions on the welcome page (there are on the extensions page and welcome page now)

it'll be make the screen easier (so no need to change something on the layout)

**Comment by cdrage** [comment]
Updated based on review, this removes portions of the "featured extensions" tests as well as they are no longer applicable / wont be showed.

The new screen looks like this:
Screenshot 2024-03-06 at 2 21 52 PM


Other PR's such as the compose onboarding screen is located here: https://github.com/containers/podman-desktop/pull/6270

**Comment by benoitf** [comment]
unit tests are failing

**Comment by cdrage** [comment]
> unit tests are failing

Was just linter, had to remove some tests since featured is no longer there, re-running now!

**Comment by benoitf** [comment]
rebasing commits on top of main branch

**Comment by benoitf** [comment]
@cdrage I tried, I'm still wondering why the checkboxes are not checked by default
I expect to go through all onboarding steps on a first run (by default)

image


also as reported previously, clicking on a card to check the item is moving the screen to the top

https://github.com/containers/podman-desktop/assets/436777/bb5676da-d19c-42a1-9c2e-c94e83c1c46c

And then I don't understand why I'm facing two screens for Podman saying it's installed

I recorded the screens
for Compose and Kubectl it's working 'as expected' , I see quickly a 'Checking ...' screen redirecting to a 'is installed' but for Podman I've got like duplicated content : I see a first 'podman is installed' screen then I click on next and I get another 'is installed' screen. I would expect the same steps than for compose and kubectl. First a 'Checking...' redirecting to the 'is installed' screen

https://github.com/containers/podman-desktop/assets/436777/1c004455-4b5b-4171-ac4e-9fbef49e772f

**Comment by cdrage** [comment]
As per call discussions, 

will remove context / when check code and instead have everything checkboxed by default.

**Comment by cdrage** [comment]
Ready for another review. There have been multiple changes so please read carefully:

- The onboarding is automatically checked now
- Changes have been made to the "double installation" page and it has now been changed to "Podman settings" which is more fitting: https://github.com/containers/podman-desktop/pull/6308
- There is a rendering bug for onboarding: https://github.com/containers/podman-desktop/issues/6313 and I have added a temporary fix here: https://github.com/containers/podman-desktop/pull/6317/commits/6e42d34c6bc944e3a165b9c776b5b4f73a9eb687
- The button that changes the height issue is located here: https://github.com/containers/podman-desktop/issues/6304 this is because the other buttons have different border's, but the "default" one has no border and  that makes it slightly smaller. This may be a longer PR as we will have to make sure that between default / primary / secondary and danger, they are all the same height.  All of them are very slightly different height due to the use of borders.

**Review by benoitf — APPROVED** [review]
I think other bugs/enhancement can be done as follow-up

probably need to ask people to try by removing their settings and uninstalling podman, compose, etc as well

**Review by deboer-tim — APPROVED** [review]

**Pull request merged**